kong/db/migrations/core/015_270_to_280.lua

```diff
@@ -0,0 +1,122 @@
+-- remove repeated targets, the older ones are not useful anymore. targets with
+-- weight 0 will be kept, as we cannot tell which were deleted and which were
+-- explicitly set as 0.
+local function c_remove_unused_targets(coordinator)
+  local cassandra = require "cassandra"
+  local upstream_targets = {}
+  for rows, err in coordinator:iterate("SELECT id, upstream_id, target, created_at FROM targets") do
+    if err then
+      return nil, err
+    end
+
+    for _, row in ipairs(rows) do
+      local key = string.format("%s:%s", row.upstream_id, row.target)
+
+      if not upstream_targets[key] then
+        upstream_targets[key] = { n = 0 }
+      end
+
+      upstream_targets[key].n = upstream_targets[key].n + 1
+      upstream_targets[key][upstream_targets[key].n] = { row.id, row.created_at }
+    end
+  end
+
+  local sort = function(a, b)
+    return a[2] > b[2]
+  end
+
+  for _, targets in pairs(upstream_targets) do
+    if targets.n > 1 then
+      table.sort(targets, sort)
+
+      for i = 2, targets.n do
+        local _, err = coordinator:execute("DELETE FROM targets WHERE id = ?", {
+          cassandra.uuid(targets[i][1])
+        })
```

Comment:
(somehow this comment was not added to my review)

I haven't tested this suggestion, but it seems to me that it's possible to delete the duplicated targets in a single loop.

```suggestion
upstream_targets[key] = {
          id = row.id, 
          created_at = row.created_at,
        }
      else
        local to_remove
        if row.created_at > upstream_targets[key].created_at then
          to_remove = upstream_targets[key].id
          upstream_targets[key] = {
            id = row.id, 
            created_at = row.created_at,
          }
        else
          to_remove = row.id
        end
        local _, err = coordinator:execute("DELETE FROM targets WHERE id = ?", {
          cassandra.uuid(to_remove)
        })
```

Reply:
Fixed and tested in my local.
It looks much simple than previous one!